go/vt/vtgate/semantics/semantic_state.go

```diff
@@ -983,3 +991,39 @@ func (st *SemTable) NewTableId() TableSet {
 	st.Tables = append(st.Tables, nil)
 	return tableID
 }
+
+func (st *SemTable) CanTakeSelectUnshardedShortcut() (*vindexes.Keyspace, bool) {
+	return canTakeSelectUnshardedShortcut(st.Tables)
+}
+
+func (st *SemTable) CanTakeUnshardedShortcut() (*vindexes.Keyspace, bool) {
+	return canTakeUnshardedShortcut(st.Tables)
+}
+
+func canTakeUnshardedShortcut(tableInfos []TableInfo) (*vindexes.Keyspace, bool) {
+	uks, _ := singleUnshardedKeyspace(tableInfos)
+	return uks, uks != nil
+}
+
+func canTakeSelectUnshardedShortcut(tableInfos []TableInfo) (*vindexes.Keyspace, bool) {
+	if mi := mirrorInfo(tableInfos); mi.Percent > 0 {
+		return nil, false
+	}
+	return canTakeUnshardedShortcut(tableInfos)
+}
+
+func (st *SemTable) GetMirrorInfo() MirrorInfo {
+	return mirrorInfo(st.Tables)
+}
+
+func mirrorInfo(tableInfos []TableInfo) MirrorInfo {
+	mi := MirrorInfo{}
+	for _, t := range tableInfos {
+		if mr := t.GetMirrorRule(); mr != nil {
+			if mi.Percent == 0 || mr.Percent < mi.Percent {
+				mi.Percent = mr.Percent
+			}
+		}
```

Comment:
we can specifically check for RealTable type here and check on the mirror rule.

Reply:
are you saying we can remove the `GetMirrorRule` method from `TableInfo` and just directly check for `t.(*RealTable)`?

Reply:
yes, we can type cast it and check the mirror rule